### PR TITLE
Product Collection: Trigger `wc-blocks_viewed_product` JS event

### LINKED ISSUE #48861: Product Collection: trigger `wc-blocks_viewed_product` JS event from product elements
> [!WARNING]  
> The issue has been repurposed. But I'm keeping the original description for context. Read the update below!

## Original issue:

**Issue Description:**
In All Products, the `experimental__woocommerce_blocks-product-view-link` action is fired when product link is clicked (it can be "View More", or "Select Options" buttons, Product Title, or Product Image). Implemented for tracking purposes for extension of [Google Analytics for WooCommerce](https://woocommerce.com/products/woocommerce-google-analytics/) plugin.

- Add to Cart Button - PC uses the Add to Cart Button in the SSR version (PHP). Block is powered by Interactivity API so the JS code that dispatches the event is not running on the front end. But there's no particular reason why it should not.
- Product Image - PC uses the Product Image in the SSR version (PHP). The JS code that dispatches the event is not running on the front end. But there's no particular reason why it should not.
- Product Title - Product Collection doesn’t use the Product Title block but the `core/post-title` variation instead. I don't think we can extend the on-click callback in the Core block. Also, I didn't investigate but for some reason, this event doesn’t even fire for All Products.

**Goal**

Implement the event so it fires from:
- Add to Cart Button
- Product Image

Consider if it can be somehow implemented in Product Title block that is `core/post-title` variation. Implementing or even finding a solution is not part of this issue.

Note: API is experimental and event payload is not documented. Still, it should be considered a public API and align with the payload provided in other places (All Products, AbstractProductGrid) as extensions rely on it.

Ref: GitHub discussion: https://github.com/woocommerce/woocommerce/discussions/48814

## Updated goal

Based on the discussion: https://github.com/woocommerce/woocommerce/discussions/48814, we decided the users of Product Collection block should rely on JS events:

I'm repurposing this issue and instead of `experimental__woocommerce_blocks-product-view-link`, the JS event `wc-blocks_viewed_product` broadcasted in a similar manner as:

- `wc-blocks_adding_to_cart` -> https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/data/cart/actions.ts/#L281-L282
- `wc-blocks_added_to_cart` -> https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/data/cart/actions.ts/#L292-L293

### PR DESCRIPTION
**Pull request by kmanijak**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds wc-blocks_viewed_product event being triggered from three places:

- Product Image
- Product Title
- Add to Cart when it's an anchor (e.g. "Select Options", "Read More", etc.)

This is the second iteration after https://github.com/woocommerce/woocommerce/pull/50217.

In opposite to the previous one, here we use a single action on the Product Collection store and trigger it from the inner blocks. Also, the Product Template provides a Interactivity context that keeps product ID.

Closes https://github.com/woocommerce/woocommerce/issues/48861

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
1. Use some plugin to include custom JS script on the page e.g. [Easy Code Snippets](https://wordpress.org/plugins/easy-code-snippets/)
2. Attach this code:

```
<script>
document.body.addEventListener(
	'wc-blocks_viewed_product',
	e => console.log('Event triggered', e.detail)
);
</script>
```

If you're using Easy Code Snippets, don't forget to wrap it in `<script></script>` tag and set the position to "Footer":
image

#### Testing
1. Make sure you have some variable, grouped and out of stock products
2. Create new post
3. Add Product Collection block
4. Choose some collection
4. Save and go to frontend
5. Open browser's console
6. Make sure you enable "Preserve log" option (example view from Chrome browser, but it may be in different place in your browser):

Screen Shot 2024-08-28 at 10 15 31 AM

8. For variable, grouped, and out-of-stock products click their "Add to Cart button" which should have labels like
    - "Select options"
    - "View products"
    - "Read more"
9. **Expected:** Along with redirection there should be an event logged in your console. Event should look like this with the `productId` corresponding to the product you clicked and `collection` corresponding to collection you chose (it's `undefined` if you chose "create your own"):

image

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts [hunk 1]
```diff
@@ -966,6 +966,63 @@ test.describe( 'Product Collection', () => {
...
+					.getByLabel( 'Select options for “V-Neck T-' )
+					.click();
+
+				const { collection, productId } = await promise;
+				expect( collection ).toEqual(
+					'woocommerce/product-collection/featured'
+				);
+				expect( productId ).toEqual( expect.any( Number ) );
+			} );
+		} );
```
Comment (gigitux) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: What do you think if we move these tests in a dedicated file like `product-collection-tracking.block_theme.spec.ts`?

**Review by gigitux — APPROVED** [review]
Great work! The PR LGTM! I left just one suggestion (but feel free to merge the PR).

Also, I have a question regarding the `viewProduct` action. I guess that we want that 3PD will adopt the same action for their inner blocks: should we document this somewhere?

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts [hunk 2]
```diff
@@ -966,6 +966,63 @@ test.describe( 'Product Collection', () => {
...
+					.locator( '[data-block-name="woocommerce/product-image"]' )
+					.nth( 0 )
+					.click();
+
+				const { collection, productId } = await promise;
+				expect( collection ).toEqual(
+					'woocommerce/product-collection/featured'
+				);
+				expect( productId ).toEqual( expect.any( Number ) );
+			} );
```
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: nit: When an event is fired after an action is performed (click in this case), we don't have to bind via `addInitScript` – we can use `page.evaluate`. For example:

```ts
const waitForViewedProductEvent = ( page ) => {
	return page.evaluate( () => {
		return new Promise( ( resolve ) => {
			window.addEventListener(
				'wc-blocks_viewed_product',
				resolve
			);
		} );
	} );
};

test( 'when Product Image is clicked', async ( { page } ) => {
	const eventPromise = waitForViewedProductEvent( page );

	await page
		.getByLabel( 'Block: Product Image' )
		.first()
		.click();

	const event = await eventPromise;

	expect( event.collection ).toEqual(
		'woocommerce/product-collection/featured'
	);
	expect( event.productId ).toEqual( expect.any( Number ) );
} );
```

This way it resembles more how we use the Playwright's `page.waitForRequest` and similar APIs. Also, if used extensively, `page.waitForEvent` could be exposed as a part of `PageUtils` fixture.

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts (hunk 1 above)
Comment (kmanijak) [reply plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: Good point! Done in https://github.com/woocommerce/woocommerce/pull/51156/commits/41c645cebe06a1d97018148e7a1c1b080fdc8d7f

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts (hunk 2 above)
Comment (kmanijak) [reply plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: I think I can't do this due to redirect that happens on click. So I'm losing a point of reference with the `page.evaluate` 🙌

**Pull request merged**